source/nodejs/adaptivecards-extras/src/progress-bar.ts

```diff
@@ -0,0 +1,117 @@
+import * as Adaptive from "adaptivecards";
+
+export class ProgressBar extends Adaptive.CardElement {
+    static readonly JsonTypeName = "Extras.ProgressBar";
+
+    //#region Schema
+
+    static readonly titleProperty = new Adaptive.StringProperty(Adaptive.Versions.v1_0, "title", true);
+    static readonly valueProperty = new Adaptive.NumProperty(Adaptive.Versions.v1_0, "value", 0);
+
+    @Adaptive.property(ProgressBar.titleProperty)
+    get title(): string | undefined {
+        return this.getValue(ProgressBar.titleProperty);
+    }
+
+    set title(value: string | undefined) {
+        if (this.title !== value) {
+            this.setValue(ProgressBar.titleProperty, value);
+
+            this.updateLayout();
+        }
+    }
+
+    @Adaptive.property(ProgressBar.valueProperty)
+    get value(): number {
+        return this.getValue(ProgressBar.valueProperty);
+    }
+
+    set value(value: number) {
+        let adjustedValue = value;
+
+        if (adjustedValue < 0) {
+            adjustedValue = 0;
+        }
+        else if (adjustedValue > 100) {
+            adjustedValue = 100;
+        }
+
+        if (this.value !== adjustedValue) {
+            this.setValue(ProgressBar.valueProperty, adjustedValue);
+
+            this.updateLayout();
+        }
+    }
+
+    //#endregion
+
+    private _titleElement?: HTMLElement;
+    private _leftBarElement: HTMLElement;
+    private _rightBarElement: HTMLElement;
+
+    protected internalRender(): HTMLElement | undefined {
+        let element = document.createElement("div");
+
+        let textBlock = new Adaptive.TextBlock();
+        textBlock.setParent(this);
+        textBlock.text = this.title;
+        textBlock.wrap = true;
+
+        this._titleElement = textBlock.render();
+
+        if (this._titleElement) {
+            this._titleElement.style.marginBottom = "6px";
+
+            let progressBarElement = document.createElement("div");
+            progressBarElement.style.display = "flex";
+
+            this._leftBarElement = document.createElement("div");
+            this._leftBarElement.style.height = "6px";
+
+            let color = Adaptive.stringToCssColor(this.hostConfig.containerStyles.emphasis.foregroundColors.accent.default);
+
+            if (color) {
+                this._leftBarElement.style.backgroundColor = color;
+            }
+
+            this._rightBarElement = document.createElement("div");
+            this._rightBarElement.style.height = "6px";
+
+            color = Adaptive.stringToCssColor(this.hostConfig.containerStyles.emphasis.backgroundColor);
```

Comment:
in general, it's a bad habit to reuse variables

Reply:
Why?

Reply:
* `color` is inadequate to describe what the variable is holding. in its first assignment, it's holding a foreground color, and in its second, a background color. this leads to cognitive load for the reader to keep in mind which color is being referred to at which point in the function
* adding new code in the vicinity of this means having to first figure out if `color`'s second life has any bearing on its first
* if this method were to be refactored, this reuse might need to be reconciled

there's really no benefit to reusing a local variable anyways

Reply:
I don't really agree with any of this, but there's also no harm in doing what you suggest. So I'll make the change if it can help closing this PR faster.

Reply:
Done

Reply:
for what it's worth, reusing variables being bad isn't just a me thing:

https://javascript.info/variables
> Reuse or create?
> And the last note. There are some lazy programmers who, instead of declaring new variables, tend to reuse existing ones.
> As a result, their variables are like boxes into which people throw different things without changing their stickers. What’s inside the box now? Who knows? We need to come closer and check.
> Such programmers save a little bit on variable declaration but lose ten times more on debugging.
> An extra variable is good, not evil.
> Modern JavaScript minifiers and browsers optimize code well enough, so it won’t create performance issues. Using different variables for different values can even help the engine optimize your code.

it also shows up in [*How To Write Unmaintainable Code*](https://cs.fit.edu/~kgallagher/Schtick/How%20To%20Write%20Unmaintainable%20Code.html), and as language checks in various places (e.g. [PHPStorm](https://www.jetbrains.com/help/phpstorm/javascript-and-typescript-reuse-of-local-variable.html))